pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/IdenticalCatchBranchesRule.java

```diff
@@ -0,0 +1,155 @@
+                if (areEquivalent(stmt, equivClass.get(0))) {
+                    equivClass.add(stmt);
+                    isNewClass = false;
+                    break;
+                }
+            }
+
+            if (isNewClass) {
+                result.add(newEquivClass(stmt));
+            }
+        }
+
+        return result;
+    }
+
+
+    private List<ASTCatchStatement> newEquivClass(ASTCatchStatement stmt) {
+        // Each equivalence class is sorted by document order
+        List<ASTCatchStatement> result = new ArrayList<>(2);
+        result.add(stmt);
+        return result;
+    }
+
+
+    // Gets the representation of the set of catch statements as a single multicatch
+    private String getCaughtExceptionsAsString(ASTCatchStatement stmt) {
+
+        StringBuilder sb = new StringBuilder();
+
+        final String delim = " | ";
+        for (ASTType type : stmt.getCaughtExceptionTypeNodes()) {
+            sb.append(type.getTypeImage()).append(delim);
+        }
+
+        // remove the last delimiter
+        sb.replace(sb.length() - 3, sb.length(), "");
+        return sb.toString();
+    }
+
+
+    @Override
+    public Object visit(ASTTryStatement node, Object data) {
+
+        List<ASTCatchStatement> catchStatements = node.getCatchStatements();
+        Set<List<ASTCatchStatement>> equivClasses = equivalenceClasses(catchStatements);
+
+        for (List<ASTCatchStatement> identicalStmts : equivClasses) {
+            if (identicalStmts.size() > 1) {
+                String identicalBranchName = getCaughtExceptionsAsString(identicalStmts.get(0));
+
+                // By convention, lower catch blocks are collapsed into the highest one
+                // The first node of the equivalence class is thus the block that should be transformed
+                for (int i = 1; i < identicalStmts.size(); i++) {
+                    addViolation(data, identicalStmts.get(i), new String[]{identicalBranchName, });
+                }
+            }
+        }
+
+        return super.visit(node, data);
+    }
+
+    /**
+     * Checks whether two nodes has same subtree.
+     * Note, it wouldn't be sensitive to changes in the exception variable name.
+     *
+     * @param node1
+     *             the first node to check
+     * @param node2
+     *             the second node to check
+     * @param exceptionName1
+     *             the first exception variable name
+     * @param exceptionName2
+     *             the second exception variable name
+     * @return <code>ture</code> if two nodes has same subtree, otherwise <code>false</code>
+     */
+    private boolean hasSameSubTree(Node node1, Node node2, String exceptionName1, String exceptionName2) {
+        if (node1 == null && node2 == null) {
+            return true;
+        } else if (node1 == null || node2 == null) {
+            return false;
+        }
+
+        if (node1.jjtGetNumChildren() == node2.jjtGetNumChildren()) {
```

Comment:
You can improve the readability of that block by reducing the nesting. Eg instead of writing 
```java
if (a) {
  // 15 lines of code
} else {
  return false; // difficult to know which `if` it referenced without a comment
}
```
You could write 
```java
if (!a) {
  return false;
}

// your 15 lines of code, which don't need to be put in an else block since the if block now ends with a return statement
```

Reply:
Done.